internal/npm_install/npm_install.bzl

```diff
@@ -469,37 +469,45 @@ def _rerooted_workspace_path(repository_ctx, f):
 def _rerooted_workspace_package_json_dir(repository_ctx):
     return str(repository_ctx.path(_rerooted_workspace_path(repository_ctx, repository_ctx.attr.package_json)).dirname)
 
-def _copy_file(repository_ctx, f):
-    to = _rerooted_workspace_path(repository_ctx, f)
-
-    # ensure the destination directory exists
-    to_segments = to.split("/")
-    if len(to_segments) > 1:
-        dirname = "/".join(to_segments[:-1])
-        args = ["mkdir", "-p", dirname] if not is_windows_os(repository_ctx) else ["cmd", "/c", "if not exist {dir} (mkdir {dir})".format(dir = dirname.replace("/", "\\"))]
-        result = repository_ctx.execute(
-            args,
-            quiet = repository_ctx.attr.quiet,
-        )
-        if result.return_code:
-            fail("mkdir -p %s failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (dirname, result.stdout, result.stderr))
+def _is_executable(repository_ctx, path):
+    stat_exe = repository_ctx.which("stat")
+    if stat_exe == None:
+        return False
+
+    # A hack to detect if stat is BSD stat as BSD stat does not support --version flag
+    is_bsd_stat = repository_ctx.execute([stat_exe, "--version"]).return_code != 0
+    if is_bsd_stat:
+        stat_args = ["-f", "%Lp", path]
+    else:
+        stat_args = ["-c", "%a", path]
 
-    # copy the file; don't use the repository_ctx.template trick with empty substitution as this
-    # does not copy over binary files properly
-    cp_args = ["cp", "-f", repository_ctx.path(f), to] if not is_windows_os(repository_ctx) else ["xcopy", "/Y", str(repository_ctx.path(f)).replace("/", "\\"), "\\".join(to_segments) + "*"]
-    result = repository_ctx.execute(
-        cp_args,
-        quiet = repository_ctx.attr.quiet,
+    arguments = [stat_exe] + stat_args
+    exec_result = repository_ctx.execute(arguments)
+    stdout = exec_result.stdout.strip()
+    mode = int(stdout, 8)
+    return mode & 0o100 != 0
+
+def _copy_file(repository_ctx, f):
+    src_path = repository_ctx.path(f)
+    dest_path = _rerooted_workspace_path(repository_ctx, f)
+    executable = _is_executable(repository_ctx, src_path)
+
+    # Copy the file
+    repository_ctx.file(
```

Comment:
something to do in a follow-up: this is very slow, so rather than reading into memory & writing do disk, we should bulk-copy inputs via `cp` in a loop

Reply:
Even better would be avoiding numerous spawns initiated via Starlark. A shell script running `cp` or a bulk copy tool (even a GulpJS stream would probably be faster).

It's on the metaphorical list.

Reply:
yep I meant running cp inside a bash for loop, not a starlark for loop